src/providers/crypto_aescbc.rs

```diff
@@ -0,0 +1,592 @@
+///
+/// Crypto provides interface fore data encryption and decryption.
+///
+/// # Examples
+/// Create cryptor with cipher key and random initialization vector.
+/// ```rust
+/// # use pubnub::{
+/// #     core::{error::PubNubError, Cryptor},
+/// #     providers::crypto_aescbc::{AesCbcIv, AesCbcCrypto}
+/// # };
+/// #
+/// # fn main() -> Result<(), PubNubError> {
+/// let cryptor = AesCbcCrypto::new("enigma", AesCbcIv::Random)?;
+/// # Ok(())
+/// # }
+/// ```
+pub struct AesCbcCrypto {
+    cipher_key: Vec<u8>,
+    iv: Option<Vec<u8>>,
+    iv_constant: bool,
+}
+
+impl AesCbcCrypto {
+    /// Create an AES-CBC crypto for data encryption and decryption.
+    ///
+    /// It is **required** that the specified initialization vector be at
+    /// maximum 16 bytes long. If you give us a value that is longer than 16
+    /// bytes, it will be clipped or filled with zeros if it is shorter.
+    ///
+    /// # Errors
+    /// Should return an [`PubNubError::CryptoInitializationError`] if cipher
+    /// key or initialization vectors are empty.
+    pub fn new<C, I>(cipher_key: C, iv: I) -> Result<Self, PubNubError>
+    where
+        C: Into<Vec<u8>>,
+        I: Into<AesCbcIv>,
+    {
+        let cipher_key: Vec<u8> = cipher_key.into();
+        let vector: AesCbcIv = iv.into();
+        let iv_constant = matches!(vector, AesCbcIv::Constant);
+
+        if let AesCbcIv::Custom(custom_iv) = &vector {
+            if custom_iv.is_empty() {
+                return Err(PubNubError::CryptoInitializationError(
+                    "Initialization vector is empty".into(),
+                ));
+            }
+        }
+
+        if cipher_key.is_empty() {
+            return Err(PubNubError::CryptoInitializationError(
+                "Cipher key is empty".into(),
+            ));
+        }
+
+        Ok(AesCbcCrypto {
+            cipher_key: AesCbcCrypto::sha256_hex(cipher_key),
+            iv: match vector {
+                AesCbcIv::Constant => Some(b"0123456789012345".to_vec()),
+                AesCbcIv::Custom(custom) => Some(custom),
+                AesCbcIv::Random => None,
+            },
+            iv_constant,
+        })
+    }
+
+    /// Calculate size of buffer for processed data.
+    ///
+    /// Buffers may require different sizes of allocated memory depending on
+    /// operation and initialization vector type.
+    fn estimated_buffer_size(&self, source: &[u8], encryption: bool) -> usize {
+        let mut size = source.len();
+
+        // More space required for encryption to align with AES cipher block
+        // size.
+        if encryption {
+            size += (AES_BLOCK_SIZE - size % AES_BLOCK_SIZE) + AES_BLOCK_SIZE;
+            // Reserve more space to store random initialization vector.
+            if !self.iv_constant {
+                size += AES_BLOCK_SIZE;
+            }
+        } else if !self.iv_constant {
+            size -= AES_BLOCK_SIZE;
+        }
```

Comment:
I'm always worried of this kind estimation of the size. 
Are this values recommended by authors of the crate?

Reply:
Okay, I see that you're using it with Vec, so there won't be any runtime errors - in worst case it will reallocate that vec.

Reply:
@Xavrax it is not a `crate` thing. It is an AES cipher block itself. There are a ways to estimate sizes when encrypting. When decrypting, I'm just decrease size on the length of IV if random has been used (as it is part of the encrypted payload).

Reply:
@Xavrax   Are you sure.. you are not gonna suggesting `match` syntax here???

Reply:
@mohitpubnub match on boolean?

Reply:
@mohitpubnub, but it didn't yell and all checks are passed ;)

Reply:
@mohitpubnub as you wish - we can go more optional syntax without mutability ;d
I like that way but I'm not sure if it is readable enough. 

maybe if move the `and_then` closure to function it will look better. 
```suggestion
        // more space required for encryption to align with aes cipher block
        // size.
        encryption
            .then_some(size + (AES_BLOCK_SIZE - size % AES_BLOCK_SIZE))
            .and_then(|encryption_size| {
                // Reserve more space to store random initialization vector.
                (!self.iv_constant)
                    .then_some(encryption_size + AES_BLOCK_SIZE)
                    .or(Some(encryption_size))
            })
            .or_else(|| (!self.iv_constant).then_some(size - AES_BLOCK_SIZE))
            .unwrap_or(size)
```

Reply:
I left it as-is so far because hard to say whether there will be really any practical benefits.
Each method name adds few bytes to compiled binary (I know that it sounds as premature optimization 😄)